File: packages/native/android/src/main/java/com/margelo/nitro/one/HybridOneNetwork.kt
package com.margelo.nitro.one

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise

// connection state matching expo-network: type plus connected and reachable
// flags from ConnectivityManager. one default-network callback is registered
// while any listener is attached: the first listener registers it, so its
// first callback is never dropped, and the last removal unregisters it.
class HybridOneNetwork : HybridOneNetworkSpec() {
    private val listeners = mutableMapOf<Int, (NetworkState) -> Unit>()
    private var nextListenerId = 0
    private var callback: ConnectivityManager.NetworkCallback? = null

    private fun connectivity(): ConnectivityManager =
        NitroModules.applicationContext?.getSystemService(Context.CONNECTIVITY_SERVICE)
            as? ConnectivityManager
            ?: throw IllegalStateException("Network: React context is not ready")

    override fun getState(): Promise<NetworkState> = Promise.async { currentState() }

    override fun addStateListener(listener: (state: NetworkState) -> Unit): () -> Unit {
        synchronized(this) {
            val id = nextListenerId++
            listeners[id] = listener
            if (callback == null) {
                val next =
                    object : ConnectivityManager.NetworkCallback() {
                        override fun onAvailable(network: Network) = emit()

                        override fun onLost(network: Network) = emit()

                        override fun onCapabilitiesChanged(
                            network: Network,
                            capabilities: NetworkCapabilities
                        ) = emit()
                    }
                callback = next
                connectivity().registerDefaultNetworkCallback(next)
            }
            return { removeListener(id) }
        }
    }

    private fun removeListener(id: Int) {
        synchronized(this) {
            listeners.remove(id)
            if (listeners.isNotEmpty()) return
            callback?.let { connectivity().unregisterNetworkCallback(it) }
            callback = null
        }
    }

    private fun emit() {
        val state = currentState()
        val current = synchronized(this) { listeners.values.toList() }
        current.forEach { it(state) }
    }

    private fun currentState(): NetworkState {
        val manager = connectivity()
        val capabilities = manager.getNetworkCapabilities(manager.activeNetwork)
            ?: return NetworkState(NetworkStateType.NONE, false, false)
        val connected = capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
        val reachable = capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)
        val type =
            when {
                !connected -> NetworkStateType.NONE
                capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> NetworkStateType.CELLULAR
                capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> NetworkStateType.WIFI
                capabilities.hasTransport(NetworkCapabilities.TRANSPORT_BLUETOOTH) -> NetworkStateType.BLUETOOTH
                capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) -> NetworkStateType.ETHERNET
                capabilities.hasTransport(NetworkCapabilities.TRANSPORT_VPN) -> NetworkStateType.VPN
                else -> NetworkStateType.OTHER
            }
        return NetworkState(type, connected, reachable)
    }
}

File: packages/native/ios/Nitro/HybridOneNetwork.swift
import Network
import NitroModules

// connection state matching expo-network: type plus connected and reachable
// flags from NWPathMonitor. one shared monitor runs while any listener is
// attached: the first listener starts it, so its first path is never
// dropped, and the last removal cancels it.
final class HybridOneNetwork: HybridOneNetworkSpec {
  private let lock = NSLock()
  private var listeners: [Int: (NetworkState) -> Void] = [:]
  private var nextListenerId = 0
  private var monitor: NWPathMonitor?

  // expo-network mapping: only a satisfied path is connected, anything else
  // is none, and reachability follows connected. ios never reports the
  // android-only other, vpn, or wimax types.
  private static func state(for path: NWPath) -> NetworkState {
    guard path.status == .satisfied else {
      return NetworkState(type: .none, isConnected: false, isInternetReachable: false)
    }
    let type: NetworkStateType
    if path.usesInterfaceType(.cellular) {
      type = .cellular
    } else if path.usesInterfaceType(.wifi) {
      type = .wifi
    } else if path.usesInterfaceType(.wiredEthernet) {
      type = .ethernet
    } else {
      type = .unknown
    }
    return NetworkState(type: type, isConnected: true, isInternetReachable: true)
  }

  func getState() throws -> Promise<NetworkState> {
    let promise = Promise<NetworkState>()
    let monitor = NWPathMonitor()
    let queue = DispatchQueue(label: "dev.onejs.network")
    // the handler and the timeout both run on this serial queue, so the flag
    // needs no lock. the first path settles; with none in five seconds the
    // read rejects instead of hanging.
    var settled = false
    monitor.pathUpdateHandler = { path in
      if settled { return }
      settled = true
      monitor.cancel()
      promise.resolve(withResult: HybridOneNetwork.state(for: path))
    }
    monitor.start(queue: queue)
    queue.asyncAfter(deadline: .now() + 5) {
      if settled { return }
      settled = true
      monitor.cancel()
      promise.reject(
        withError: oneNativeError(
          "E_NETWORK_TIMEOUT", "Network.getState: timed out waiting for the network path."))
    }
    return promise
  }

  func addStateListener(listener: @escaping (_ state: NetworkState) -> Void) throws -> () -> Void {
    lock.lock()
    defer { lock.unlock() }
    let id = nextListenerId
    nextListenerId += 1
    listeners[id] = listener
    if monitor == nil {
      let next = NWPathMonitor()
      next.pathUpdateHandler = { [weak self] path in
        self?.emit(HybridOneNetwork.state(for: path))
      }
      next.start(queue: .main)
      monitor = next
    }
    return { [weak self] in self?.removeListener(id) }
  }

  private func emit(_ state: NetworkState) {
    lock.lock()
    let current = Array(listeners.values)
    lock.unlock()
    for listener in current {
      listener(state)
    }
  }

  private func removeListener(_ id: Int) {
    lock.lock()
    defer { lock.unlock() }
    listeners[id] = nil
    if listeners.isEmpty {
      monitor?.cancel()
      monitor = nil
    }
  }
}
